crates/bevy_ecs/src/bundle.rs

```diff
@@ -168,46 +190,37 @@ pub unsafe trait Bundle: DynamicBundle + Send + Sync + 'static {
         // Ensure that the `OwningPtr` is used correctly
         F: for<'a> FnMut(&'a mut T) -> OwningPtr<'a>,
         Self: Sized;
-
-    /// Registers components that are required by the components in this [`Bundle`].
-    fn register_required_components(
-        _components: &mut Components,
-        _required_components: &mut RequiredComponents,
-    );
 }
 
 /// The parts from [`Bundle`] that don't require statically knowing the components of the bundle.
 pub trait DynamicBundle {
+    /// An operation on the entity that happens _after_ inserting this bundle.
+    type Effect: BundleEffect;
     // SAFETY:
     // The `StorageType` argument passed into [`Bundle::get_components`] must be correct for the
     // component being fetched.
     //
     /// Calls `func` on each value, in the order of this bundle's [`Component`]s. This passes
     /// ownership of the component values to `func`.
     #[doc(hidden)]
-    fn get_components(self, func: &mut impl FnMut(StorageType, OwningPtr<'_>));
+    fn get_components(self, func: &mut impl FnMut(StorageType, OwningPtr<'_>)) -> Self::Effect;
+}
+
+/// An operation on an [`Entity`] that occurs _after_ inserting the [`Bundle`] that defined this bundle effect.
```

Comment:
I think the doc should specify how the Bundle effect is ordered with respect to Hooks/Observers that the bundle will trigger

Reply:
Yeah thats a good call.

> to immediately flush commands and apply the hook/observer as we spawn a Bundle

You can definitely do that, although its worth noting that we already apply hooks / observers on spawn now (this was recently added)